pkg/gornir/gornir.go

```diff
@@ -16,6 +17,77 @@ type Gornir struct {
 	Logger    Logger     // Logger for the object
 }
 
+// InventoryPlugin is an Inventory Source
+type InventoryPlugin interface {
+	Create() (Inventory, error)
+}
+
+// New is a Gornir constructor. It is currently no different that new,
+// however is a placeholder for any future defaults.
+func New() *Gornir {
+	return new(Gornir)
+}
+
+// SetOption is a funcion that sets one or more options for a given Gornir.
+type SetOption func(r *Gornir) error
+
+// Build construct a new Gornir from scratch with options.
+func Build(opts ...SetOption) (*Gornir, error) {
+	var gornir Gornir
+	for _, opt := range opts {
+		err := opt(&gornir)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &gornir, nil
+}
+
+// WithInventory reads the inventory from a file for a Gornir.
+func WithInventory(p InventoryPlugin) SetOption {
+	return func(g *Gornir) error {
+		inv, err := p.Create()
+		if err != nil {
+			return errors.Wrap(err, "could not read inventory from plugin")
+		}
+		g.Inventory = &inv
+		return nil
+	}
+}
+
+// WithLogger sets the logging option for a Gornir.
+func WithLogger(l Logger) SetOption {
+	return func(g *Gornir) error {
+		if l == nil {
+			errors.New("didn't receive a valid logger")
+		}
+		g.Logger = l
+		return nil
+	}
+}
+
+// WithFilter provides a FilterFunc to a Gornir to filter the list of hosts.
+func WithFilter(f FilterFunc) SetOption {
+	return func(g *Gornir) error {
+		if f == nil {
+			errors.New("didn't receive a valid filter function")
```

Comment:
missing return

Reply:
It returns a SetOption function: `return func(g *Gornir) error{...}`

Reply:
> Awesome work :)
> 
> Give me a couple of days for a more thorough review though (local holiday here tomorrow)

No rush, this gives me time to re-think this.